core/src/schema/content/array.rs

```diff
@@ -1,15 +1,138 @@
+                                return Err(de::Error::duplicate_field("length"));
+                            }
+
+                            length = Some(access.next_value()?);
+                        }
+                        Field::Content => {
+                            if content.is_some() {
+                                return Err(de::Error::duplicate_field("content"));
+                            }
+
+                            content = Some(access.next_value()?);
+                        }
+                    }
+                }
+
+                let length = length.ok_or_else(|| de::Error::missing_field("length"))?;
+                let content = content.ok_or_else(|| de::Error::missing_field("content"))?;
+
+                match length {
+                    Content::Number(NumberContent::U64(number_content::U64::Range(r))) => {
+                        if r.high.is_none() {
+                            return Err(A::Error::custom(
+                                "missing high value for array length range"
+                            ));
+                        }
+                    }
+                    // Default for ranges
+                    Content::Number(NumberContent::U32(number_content::U32::Range(r))) => {
+                        if r.high.is_none() {
+                            return Err(A::Error::custom(
+                                "missing high value for array length range"
+                            ));
+                        }
+                    }
+                    Content::Number(NumberContent::U64(_)) => {},
+                    // Default for negative numbers
+                    Content::Number(NumberContent::I64(number_content::I64::Constant(n))) => {
+                        is_positive(n).map_err(A::Error::custom)?
+                    }
+                    Content::Number(NumberContent::I32(number_content::I32::Range(r))) => {
+                        if r.high.is_none() {
+                            return Err(A::Error::custom(
+                                "missing high value for array length range"
+                            ));
+                        }
+
+                        is_positive(
+                            r.low .unwrap_or_default() .into(),
+                        ).map_err(A::Error::custom)?
+                    },
+                    Content::SameAs(_) => {},
+                    Content::Null(_) => return Err(de::Error::custom("array length is missing. Try adding '\"length\": 5' to the array type")),
+                    Content::Empty(_) => return Err(de::Error::custom("array length is not a constant or number type. Try replacing the '\"length\": {}' with '\"length\": [number]' where '[number]' is a positive integer")),
+                    Content::OneOf(ref one) => if one.variants.iter().any(|variant| variant == &*NULL_VARIANT) {
+                        return Err(de::Error::custom("cannot use '\"optional\": true' in array length"));
```

Comment:
Not to be too nitpicky again, but users could specify `one_of` generator that has a `"type": "null"` generator in the `"variants": [...]` list - perhaps the message could be changed to something like `"cannot use 'one_of' generator nor '"optional": true' in array length"?

Or perhaps it would make more sense to add a field to `OneOfContent` that indicates whether or not is was specified using the `"optional": true` method or manually as `{ "type": "one_of", ... }` - then your existing error messages could be used and `if one.variants.iter().any(|variant| variant == &*NULL_VARIANT) { ...` could be replaced with `if one.specified_with_optional_true { ...` or similar.

Reply:
Not nitpicky at all. I still have blind spots and didn't realize `optional` makes a `one_of` with a `null` variant internally. So this helps a lot :smile: